compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignNpuDmaBdIds.cpp

```diff
@@ -158,8 +262,18 @@ LogicalResult assignNpuDmaBdIds(AMDAIE::WorkgroupOp workgroupOp) {
         }
         ChannelBdIdGenerator &generator =
             shimTileToGeneratorMap[tileOp.getResult()];
-        uint32_t value = bdIdOp.getValue();
-        generator.releaseBdId(value);
+        Value value = bdIdOp.getValue();
+        if (auto addOp = value.getDefiningOp<arith::AddIOp>()) {
+          // If the BD ID is a semi-affine expression, retrieve the BD ID for
+          // the first iteration.
+          FailureOr<uint32_t> bdId = retriveBdId(addOp);
+          if (failed(bdId)) return WalkResult::interrupt();
+          generator.releaseBdId(*bdId);
```

Comment:
Shouldn't all of them be released?

Reply:
Assume the number of DMAs per iteration of the loop is even, while the total number of available BD IDs is odd. In this case, the ownership of BD IDs might get interleaved. Since the logic here was assigning on DMA copy and releasing on DMA wait, it can't handle this "complex" case. As a hacky approach, I only assign / release BD ID for the first iteration. It's probably fine if it is the innermost loop, but I will think about a better solution.

Reply:
> Assume the number of DMAs per iteration of the loop is even, while the total number of available BD IDs is odd. In this case, the ownership of BD IDs might get interleaved.

Could you show this case with an example? I would expect every `dma_cpy_nd` to have a fixed range of BD ids assigned for the loop iterations, so if `nb_dma_cpy_nd_ops * nb_bd_ids_per_dma_cpy_nd < nb_available_bd_ids`, there shouldn't be an issue?

I think annotating the full range of BD ids as assigned should be a safe thing to do, even with nested inner loops:

```
for iv = 0 to 4 step 1 {
  %0 = dma_cpy_nd(bd_id = iv)  // simplified
  %1 = dma_cpy_nd(bd_id = 4 +  iv)  // simplified
  for iv2 = 0 to 2 step 1 {
    %2 = dma_cpy_nd(bd_id = 8 +  iv2)  // simplified
    dma_wait(%2)
  }
  dma_wait(%0)
  dma_wait(%1)
}
```

Here, the most nested loop can't use BD ranges [0, 4[ and [4, 8[ because they are already used.

As the wait ops release the full range of BD ids, this works as well with:

```
for iv = 0 to 4 step 1 {
  %0 = dma_cpy_nd(bd_id = iv)  // simplified
  %1 = dma_cpy_nd(bd_id = 4 +  iv)  // simplified
  dma_wait(%0)
  dma_wait(%1)
  for iv2 = 0 to 2 step 1 {
    %2 = dma_cpy_nd(bd_id = iv2)  // simplified
    dma_wait(%2)
  }
}
```
Here, the BD ranges [0, 4[ and [4, 8[ are released before the inner loop, so they can be reused.

Reply:
I think there are currently two ways:

The main difference might be when number of available BD is not divisible by number of DMA. For Method I, I guess the intention is also to leave some BD never utilized. For example, if the allowed BD range is [0,8], 8 is never used. I am using Method II for now, but the issue is this statement is no longer true.... `I would expect every dma_cpy_nd to have a fixed range of BD ids assigned for the loop iterations,`

Method I:
```
for iv = 0 to 64 step 1 {
  %0 = dma_cpy_nd(bd_id = iv mod 8)            // get 0, 1, 2, 3,   0, 1, 2, 3
  %1 = dma_cpy_nd(bd_id = (4 +  iv) mod 8)     // get 4, 5, 6, 7,    4, 5, 6, 7
}
```

Method II:
```
for iv = 0 to 64 step 1 {
  %0 = dma_cpy_nd(bd_id = (2*iv) mod 9)      // get 0, 2, 4, 6, 8,    1, 3, 5, 7 ....
  %1 = dma_cpy_nd(bd_id = (2*iv+1) mod 9)    // get 1, 3, 5, 7, 0,    2, 4, 6, 8 ....
}
```

Reply:
Now I am thinking Method I is probably better, given the cost is small (losing 1 BD in this case)

Reply:
> I guess the intention is also to leave some BD never utilized. For example, if the allowed BD range is [0,8], 8 is never used. 

That's true in this example, but it doesn't have a performance impact and it might be useful to have this additional BD in other cases:

```
%0 = dma_cpy_nd(bd_id = 0)  
for iv = 0 to 64 step 1 {
  %1 = dma_cpy_nd(bd_id = 1 + iv mod 8)            // get 1, 2, 3, 4,   1, 2, 3, 4
  %2 = dma_cpy_nd(bd_id = 1+  (4 +  iv) mod 8)     // get 5, 6, 7, 8,   5, 6, 7, 8
}
```

```
for iv = 0 to 64 step 1 {
  %0 = dma_cpy_nd(bd_id = iv mod 8)            // get 0, 1, 2, 3,   0, 1, 2, 3
  %1 = dma_cpy_nd(bd_id = (4 +  iv) mod 8)     // get 4, 5, 6, 7,    4, 5, 6, 7
  %2 = dma_cpy_nd(bd_id = 8)                        // always 8, even though we can't use different BD ids and we won't be able to chain, it's still a valid solution
}
```

Reply:
And for the nested loop, I haven't seen any actual example has this pattern
```ll
for iv = 0 to 4 step 1 {
  %0 = dma_cpy_nd(bd_id = iv)  // simplified
  %1 = dma_cpy_nd(bd_id = 4 +  iv)  // simplified
  dma_wait(%0)
  dma_wait(%1)
  for iv2 = 0 to 2 step 1 {
    %2 = dma_cpy_nd(bd_id = iv2)  // simplified
    dma_wait(%2)
  }
}
```

Instead, this happens a lot
```ll
for iv = 0 to 4 step 1 {
  %0 = dma_cpy_nd(bd_id = iv)  // simplified
  %1 = dma_cpy_nd(bd_id = 4 +  iv)  // simplified
  for iv2 = 0 to 2 step 1 {
    %2 = dma_cpy_nd(bd_id = iv2)  // simplified
    dma_wait(%2)
  }
  dma_wait(%0)
  dma_wait(%1)
}
```
Theoretically, for the second case, we can still have the semi-affine expression for the outer loop, as long as (inner loop iteration * inner loop per iteration DMA) < available BD. 

Given that, maybe I can stay with the constraint about InnerMostLoop, and add a TODO there?

Reply:
Maybe we haven't seen the first example

````

> And for the nested loop, I haven't seen any actual example has this pattern
> 
> ```llvm
> for iv = 0 to 4 step 1 {
>   %0 = dma_cpy_nd(bd_id = iv)  // simplified
>   %1 = dma_cpy_nd(bd_id = 4 +  iv)  // simplified
>   dma_wait(%0)
>   dma_wait(%1)
>   for iv2 = 0 to 2 step 1 {
>     %2 = dma_cpy_nd(bd_id = iv2)  // simplified
>     dma_wait(%2)
>   }
> }
> ```

Maybe we haven't seen this, but it's valid IR that could come up, and I don't see the issue in supporting it? I think that if you assign the range of values used for a loop when you encounter the `dma_cpy_nd` and release them on wait, it should handle both cases above, or am I missing something?

Reply:
Yeah, so I guess when encountering the `dma_cpy_nd`, just adding some extra logic to check whether `dma_wait` is before/after any innerloop. And how about the case of wait after innerloop, shall we also try to have expression if possible or leave it for now?

Reply:
> Yeah, so I guess when encountering the dma_cpy_nd, just adding some extra logic to check whether dma_wait is before/after any innerloop.

Why would you need that additional logic? I think you can loop through the range of values when you encounter `dma_cpy_nd` and assign them/annotate them as assigned and then release all of them when you encounter `dma_wait`. I think this should avoid needing special logic for innermost, wait before/after inner loop.

Reply:
Because when encountering `dma_cpy_nd`, I need to decide beforehand whether to use BD as a constant or an expression. If it's an expression, I also need to determine its range, as I want to avoid leaving the inner loop without any BD to assign. However, all of this depends on the location of `dma_wait`.

Reply:
> Because when encountering dma_cpy_nd, I need to decide beforehand whether to use BD as a constant or an expression. If it's an expression, I also need to determine its range, as I want to avoid leaving the inner loop without any BD to assign. However, all of this depends on the location of dma_wait.

You could traverse the controlcode from inner to outer?

Reply:
Potentially yes, but the BD release might be getting tricky. When the inner loop releases BDs, the outer loop actually needs avoid using them.